### ISSUE TITLE
Support for virtual classes?

### ISSUE THREAD

**Issue by hadley** [RConsortium/OOP-WG#41]
Do we expect the system to support virtual classes?

**Comment by lawremi** [comment]
Good question. With S4, they've proven useful for documenting the role of a class, but that value is fairly limited. Obviously, in a dynamically typed language, we rely on duck typing to circumvent the need for abstract/virtual methods that would necessitate an abstract/virtual class.

That said, they would be simple to implement.

**Comment by mmaechler** [comment]
Indeed, in S4.  For me, not only useful for documenting, but IIRC quite crucial in building a logical class hierarchy inheriting slots, and allowing the DRY principle.  E.g. (from memory; may be wrong in the details): a `"Matrix"` has a `Dim` property (which must be numeric (non-negative, no NA) of length 2) but "Matrix" itself is virtual, as an "actual" (my nomenclature of "non-virtual") Matrix needs more structure (e.g. "numeric" vs "logical" in its content).

**Comment by wlandau** [comment]
Is there anything in particular we would need to do? To me, the ability to define a virtual/abstract class seems like a natural consequence of inheritance. I use them in S3 and R6 quite heavily in [`targets`](https://github.com/wlandau/targets), and I find them super helpful.

**Comment by ltierney** [comment]
I'm not sure I see anything special is needed. The usual idea is that virtual classes aren't supposed to be directly instantiated, but that can be handled by convention rather than needing an enforcement mechanism.

**Comment by hadley** [comment]
Sounds like the consensus is that virtual classes are useful and easy to implement, so I'll add to the spec.

**Comment by lawremi** [comment]
I guess @ltierney and I are not entirely convinced we need to formally support them, but yes, they would be easy to implement.
